src/macaron/parsers/bashparser.py

```diff
@@ -115,78 +148,95 @@ def parse(bash_content: str, macaron_path: str = "") -> dict:
+    repo_path: str,
+    caller: BaseNode,
+    recursion_depth: int,
+    macaron_path: str | None = None,
+) -> BashNode:
+    """Create a callgraph node for a bash script.
+
+    A bash node can have the following types:
+
+      * :class:`BashScriptType.INLINE` when it is inlined in a CI workflow.
+      * :class:`BashScriptType.FILE` when it is a bash script file.
 
     Parameters
     ----------
-    bash_content : str
-        Bash script content.
-    ci_file : str
-        The relative path to the entry point CI workflow.
-    ci_type : str
-        The CI service.
-    macaron_path : str
-        Macaron's root path (optional).
-    recursive : bool
-        Recursively parse bash scripts with depth 1.
-        Should specify repo_path too if set to True.
-    repo_path : str
+    name: str
+        A name to be used as the identifier of the node.
+    node_type: BashScriptType
+        The type of the node.
+    source_path: str
+        The file that contains the bash script.
+    parsed_obj: dict | None
+        The parsed bash scrit object.
+    repo_path: str
         The path to the target repo.
-    working_dir : str
-        The working directory from which the bash script has run.
-        Empty value is considered as the root of the repo.
-    job_name: str
-        The name of the job where commands were called.
-    step_name: str
-        The name of the step where commands were called.
-
-    Yields
+    caller: BaseNode
+        The caller node.
+    recursion_depth: int
+        The number of times this function is called recursively.
+    macaron_path=None
+        The path to the Macaron module.
+
+    Returns
+    -------
+    BashNode
+        A bash node object.
+
+    Raises
     ------
-    BashCommands
-        The parsed bash script objects.
+    CallGraphError
+        When unable to create a bash node.
     """
-    if not macaron_path:
-        macaron_path = global_config.macaron_path
-
-    parsed_parent = parse(bash_content)
-    caller_commands = parsed_parent.get("commands", [])
-    if caller_commands:
-        yield BashCommands(
-            caller_path=ci_file,
-            CI_path=ci_file,
-            CI_type=ci_type,
-            commands=caller_commands,
-            job_name=job_name,
-            step_name=step_name,
-        )
+    if recursion_depth > defaults.getint("bashparser", "recursion_depth", fallback=3):
+        raise CallGraphError(f"The analysis has reached maximum recursion depth {recursion_depth} at {source_path}.")
+    parsed_content = {}
+    working_dir = None
+    match node_type:
+        case BashScriptType.INLINE:
+            if parsed_obj is None:
+                raise CallGraphError(f"Unable to find the parsed AST for the CI step at {source_path}.")
+            step_exec = validate_step(parsed_obj)
+            if step_exec is None:
+                raise CallGraphError(f"Unable to validate parsed AST for the CI step at {source_path}.")
```

Comment:
I think because `validate_step` perform the validation and returns None if the validation failed, we can change the content of this exception to
```suggestion
                raise CallGraphError(f"Validation failed for the parsed AST for the CI step at {source_path}.")
```

Reply:
I don't really see any difference in meaning between the two.

Reply:
Hmm, I see. Then it's fine for me to leave it as it is.

Reply:
Resolved due to being similar to https://github.com/oracle/macaron/pull/673#discussion_r1535073849